### ISSUE TITLE
remove/add entry sequence for zip archives created with macOS Finder

### ISSUE THREAD

**Issue by potatobeanradio** [weichsel/ZIPFoundation#238]
# Summary
When performing a remove/addEntry sequence of operations on an Archive, the new entry does not appear in the archive if the original .zip file was created with the macOS finder. 

# Steps to Reproduce

command line program below reproduces the issue when used with attached .zip files:
_______

```swift
import Foundation
import ZIPFoundation


testArchive(fileURL: URL(fileURLWithPath: "ZIPFoundationArchive.zip"))
testArchive(fileURL: URL(fileURLWithPath: "Archive.zip"))

func testArchive(fileURL: URL) {
    
    
    print("\n -----Testing remove/add for: \(fileURL.path) ------")

    let archive = Archive(url: fileURL, accessMode: .update)!

    let entryName = "myFile.txt"

    if let entry = archive[entryName] {
        try! archive.remove(entry)
        print("\(entry.path) removed from the archive")
    } else {
        print("could not find: \(entryName)")
    }


    let newData = "my data string".data(using: .utf8)!
    try! archive.addEntry(with: entryName, type: .file, uncompressedSize: Int64(newData.count), provider: {position, size in
        return newData.subdata(in: Int(position)..<Int(position)+size)
    })
    print("added entry: \(entryName)")

    if let entry = archive[entryName] {
        print("\(entry.path) is found in archive")
    } else {
        print("Error!  could not find: \(entryName) despite adding it to archive")
    }

}
```

_____
Note:  Archive.zip was created in the macOS 12.0.1 Finder using the 'Compress' contextmenu.  ZIPFoundationArchive.zip was created by programmatically using ZIPFoundation itself. 

[Archive.zip](https://github.com/weichsel/ZIPFoundation/files/7663607/Archive.zip)

[ZIPFoundationArchive.zip](https://github.com/weichsel/ZIPFoundation/files/7663608/ZIPFoundationArchive.zip)

# Expected Results
myFile.txt is found in archive for both ZIPFoundation.zip and Archive.zip

# Actual Results
myFile.txt is not found in archive Archive.zip

# Regression & Version
Xcode 13.1
macOS 12.0.1
ZIPFoundation 0.9.13

# Related Link

**Comment by weichsel** [comment]
Thanks for the report - I can reproduce the problem with ZIPFoundation 0.9.13.

Looks like a regression with 0.9.13 - the remove/add sequence works properly with 0.9.12. 

I haven't investigated in detail yet but the reason why only the Finder-generated archive is affected could be due to the entry count being > 1 (there's a hidden `.DS_Store` directory in there).
Maybe we are mis-calculating some offsets in `remove(entry: ...)`? /cc @Ckitakishi

**Comment by weichsel** [comment]
I think I found the problem. With the latest release, the "Extra Field Data" (which is optional and can contain all sorts of  things such as time zones for modifications dates etc.) was not preserved properly for the remaining entries when an entry got removed. 
While this doesn't cause any data integrity issues, it can lead to wrong offsets in some ZIP headers. Pedantic programs or libraries (like ZIPFoundation 🙈) might return `nil` when encountering mismatching offsets during entry lookup. 

The issue should be fixed in the above PR. Will test a bit more and then tag a new release.
